### PR TITLE
dylib-unhell.py: simplify process_libraries() to eliminate leafs()

### PR DESCRIPTION
**Pull request by downthomas**
Instead of traversing across leafs() which can lead to an infinite loop issue with cross-linked libraries, apply the set() function to the dictionary created by libraries() to create a set (libs_set) without duplicates.
Remove binary from libs_set so that a duplicate of the binary is not added to the libs directory.
Iterate over libs_set to bundle dylibs while using the dictionary (libs_dict) to determine which install_names to change.

I tested this on Mac OS 10.14 with both python 2.7.16 and 3.8.0. Before this edit, I experienced an infinite loop issue because I have freetype and harfbuzz crosslinked. After this edit, dylib-unhell.py worked and exited normally.

I offer these changes into the public domain so they can be used for any purpose whatsoever.

### CODE REVIEW

**Comment by Akemi** [comment]
i am not quite sure this works out as intended, but i need to look into this a bit more when i get some time.

**Comment by downthomas** [comment]
I stumbled upon this solution by accident, so let me walk you through how it works. 
The `libraries()` function returns a dictionary (`libs_dict`) with all the libraries linked to mpv, and all the libraries linked to those libraries (and any libraries linked to those libraries, ad infinitum). Each of those libraries is a key in `libs_dict`. For example:

```
libs_dict:
{'/Volumes/RAMDisk/mpv/build/mpv.app/Contents/MacOS/mpv': {'/usr/local/lib/libavcodec.58.dylib', '/usr/local/lib/libluajit-5.1.2.dylib', '/usr/local/lib/libarchive.13.dylib', '/usr/local/lib/libavdevice.58.dylib', '/usr/local/lib/libavutil.56.dylib', '/usr/local/lib/libuchardet.0.dylib', '/usr/local/lib/libavformat.58.dylib', '/usr/local/lib/libswscale.5.dylib', '/usr/local/lib/libswresample.3.dylib', '/usr/local/lib/libass.9.dylib', '/usr/local/lib/libvapoursynth-script.0.dylib', '/usr/local/lib/libjpeg.7.dylib', '/usr/local/lib/liblcms2.2.dylib', '/usr/local/lib/libavfilter.7.dylib'},
 '/usr/local/lib/libavcodec.58.dylib': {'/usr/local/lib/libwebp.7.dylib', '/usr/local/lib/libx264.158.dylib', '/usr/local/lib/libavutil.56.dylib', '/usr/local/lib/liblzma.5.dylib', '/usr/local/lib/libopus.0.dylib', '/usr/local/lib/libswresample.3.dylib', '/usr/local/lib/libvpx.6.dylib', '/usr/local/lib/libx265.184.dylib'},
 '/usr/local/lib/libwebp.7.dylib': set(),
 '/usr/local/lib/libx264.158.dylib': set(),
 '/usr/local/lib/libavutil.56.dylib': set(),
 '/usr/local/lib/liblzma.5.dylib': set(),
 '/usr/local/lib/libopus.0.dylib': set(),
 '/usr/local/lib/libswresample.3.dylib': {'/usr/local/lib/libavutil.56.dylib'},
 '/usr/local/lib/libvpx.6.dylib': set(),
 '/usr/local/lib/libx265.184.dylib': set(),
 '/usr/local/lib/libluajit-5.1.2.dylib': set(),
 '/usr/local/lib/libarchive.13.dylib': {'/usr/local/lib/libnettle.7.dylib', '/usr/local/lib/liblzma.5.dylib'},
 '/usr/local/lib/libnettle.7.dylib': set(),
 '/usr/local/lib/libavdevice.58.dylib': {'/usr/local/lib/libavformat.58.dylib', '/usr/local/lib/libavcodec.58.dylib', '/usr/local/lib/libavfilter.7.dylib'},
 '/usr/local/lib/libavformat.58.dylib': {'/usr/local/lib/libgmp.10.dylib', '/usr/local/lib/libavutil.56.dylib', '/usr/local/lib/libavcodec.58.dylib'},
 '/usr/local/lib/libgmp.10.dylib': set(),
 '/usr/local/lib/libavfilter.7.dylib': {'/usr/local/lib/libfreetype.6.dylib', '/usr/local/lib/libpostproc.55.dylib', '/usr/local/lib/libavcodec.58.dylib', '/usr/local/lib/libavutil.56.dylib', '/usr/local/lib/libavformat.58.dylib', '/usr/local/lib/libswscale.5.dylib', '/usr/local/lib/libswresample.3.dylib', '/usr/local/lib/libass.9.dylib', '/usr/local/lib/libfribidi.0.dylib'},
 '/usr/local/lib/libfreetype.6.dylib': {'/usr/local/lib/libpng16.16.dylib', '/usr/local/lib/libharfbuzz.0.dylib'},
 '/usr/local/lib/libpng16.16.dylib': set(),
 '/usr/local/lib/libharfbuzz.0.dylib': {'/usr/local/lib/libfreetype.6.dylib'},
 '/usr/local/lib/libpostproc.55.dylib': {'/usr/local/lib/libavutil.56.dylib'},
 '/usr/local/lib/libswscale.5.dylib': {'/usr/local/lib/libavutil.56.dylib'},
 '/usr/local/lib/libass.9.dylib': {'/usr/local/lib/libfreetype.6.dylib', '/usr/local/lib/libfribidi.0.dylib', '/usr/local/lib/libharfbuzz.0.dylib'},
 '/usr/local/lib/libfribidi.0.dylib': set(),
 '/usr/local/lib/libuchardet.0.dylib': set(),
 '/usr/local/lib/libvapoursynth-script.0.dylib': set(),
 '/usr/local/lib/libjpeg.7.dylib': set(),
 '/usr/local/lib/liblcms2.2.dylib': set()}
```
Using set() on a dictionary will return the keys of the dictionary. So in this case:
```
libs_set: 
{'/usr/local/lib/libfreetype.6.dylib',
 '/usr/local/lib/libpostproc.55.dylib',
 '/usr/local/lib/libavcodec.58.dylib',
 '/usr/local/lib/libpng16.16.dylib',
 '/usr/local/lib/libarchive.13.dylib',
 '/Volumes/RAMDisk/mpv/build/mpv.app/Contents/MacOS/mpv',
 '/usr/local/lib/libharfbuzz.0.dylib',
 '/usr/local/lib/libuchardet.0.dylib',
 '/usr/local/lib/liblzma.5.dylib',
 '/usr/local/lib/libavformat.58.dylib',
 '/usr/local/lib/libswscale.5.dylib',
 '/usr/local/lib/libswresample.3.dylib',
 '/usr/local/lib/libass.9.dylib',
 '/usr/local/lib/libvpx.6.dylib',
 '/usr/local/lib/libvapoursynth-script.0.dylib',
 '/usr/local/lib/libx265.184.dylib',
 '/usr/local/lib/libfribidi.0.dylib',
 '/usr/local/lib/libx264.158.dylib',
 '/usr/local/lib/libluajit-5.1.2.dylib',
 '/usr/local/lib/libavutil.56.dylib',
 '/usr/local/lib/libavdevice.58.dylib',
 '/usr/local/lib/libnettle.7.dylib',
 '/usr/local/lib/libopus.0.dylib',
 '/usr/local/lib/libgmp.10.dylib',
 '/usr/local/lib/libjpeg.7.dylib',
 '/usr/local/lib/libwebp.7.dylib',
 '/usr/local/lib/liblcms2.2.dylib',
 '/usr/local/lib/libavfilter.7.dylib'}
```

We can be sure there will be no duplicates in the set [because `libraries()` checks to see if a library in `libs_list` is in the `result`,](https://github.com/mpv-player/mpv/blob/b3b2cc44fa783c1575b3e229c79f2977f75a57c1/TOOLS/dylib-unhell.py#L58-L60). If it isn't, it calls `libraries()` on the lib to add it to `result`, which ultimately becomes `libs_dict`. Therefore, every non-empty set value in libs_dict will also be a key in libs_dict, so we can know with certainty that every unique library is a key in libs_dict. So we just need the keys from libs_dict, which is what we get with libs_set. 

From there the code is unchanged except for change `for p in libs_set`, because that iterates over every unique library and checks to see if it is also a value in libs_dict for `install_name_tool_change`.

Let me know if you have any questions, and I'd encourage anyone reading this to test it out and give any feedback.

File: TOOLS/dylib-unhell.py [hunk 1]
```diff
@@ -94,11 +84,11 @@ def process_libraries(libs_dict, binary, processed = []):
         if src in libs_dict[binary]:
             install_name_tool_change(src, name, binary)
 
-        for p in processed:
+        for p in libs_set:
             if p in libs_dict[src]:
                 install_name_tool_change(p, lib_name(p), dst)
 
-    return ls
+    return
```
Comment (Akemi) [inline TOOLS/dylib-unhell.py]: this can be removed

**Comment by Akemi** [comment]
i would prefer the commit prefix "osxbundle:" for consistency. also from the contribute.md the commit message body and head should not be longer than 72 characters.

otherwise this looks good and still does what is intended.

File: TOOLS/dylib-unhell.py (hunk 1 above)
Comment (downthomas) [reply TOOLS/dylib-unhell.py]: Removed in 78161279713c9ea766967545ba482cc5017083b2.

**Comment by downthomas** [comment]
Thanks. Sorry about not following the formatting requirements, that was my mistake.
I amended the commit subject for consistency, and amended the commit message for both clarity and formatting. I also added a commit to remove the unnecessary return.

**Comment by Akemi** [comment]
if you could merge both commits together that would be nice too. otherwise i would do it tomorrow when i am back home.

**Comment by downthomas** [comment]
Merged. :)

**Comment by Akemi** [comment]
thanks for you work.

**Pull request merged**